File: ruby/core/test/mullet/default_scope_test.rb
require 'minitest/autorun'
require 'mullet/default_scope'

class DefaultScopeTest < MiniTest::Unit::TestCase
  VARIABLE_NAME = :first_name
  VARIABLE_VALUE = 'Joe'

  def test_return_this_object()
    data = VARIABLE_VALUE
    scope = Mullet::DefaultScope.new(data)
    assert_equal VARIABLE_VALUE, scope.get_variable_value(:'.')
  end

  def test_return_value_from_hash()
    data = { VARIABLE_NAME => VARIABLE_VALUE }
    scope = Mullet::DefaultScope.new(data)
    assert_equal VARIABLE_VALUE, scope.get_variable_value(VARIABLE_NAME)
  end

  Struct.new('Person', VARIABLE_NAME)

  def test_return_method_return_value()
    data = Struct::Person.new(VARIABLE_VALUE)
    scope = Mullet::DefaultScope.new(data)
    assert_equal VARIABLE_VALUE, scope.get_variable_value(VARIABLE_NAME)
  end

  class Person
    def initialize(first_name)
      @first_name = first_name
    end
  end

  def test_return_instance_variable_value()
    data = Person.new(VARIABLE_VALUE)
    scope = Mullet::DefaultScope.new(data)
    assert_equal VARIABLE_VALUE, scope.get_variable_value(VARIABLE_NAME)
  end

  def test_return_proc_return_value()
    variable_proc = Proc.new { VARIABLE_VALUE }
    data = Person.new(variable_proc)
    scope = Mullet::DefaultScope.new(data)
    assert_equal VARIABLE_VALUE, scope.get_variable_value(VARIABLE_NAME)
  end
end

File: java/core/src/test/java/com/github/pukkaone/mullet/DefaultScopeTest.java
/*
Copyright (c) 2011, Chin Huang
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice,
   this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package com.github.pukkaone.mullet;

import static org.junit.Assert.assertEquals;

import java.util.Collections;
import java.util.concurrent.Callable;
import org.junit.Test;

public class DefaultScopeTest {

    private static final String VARIABLE_NAME = "firstName";
    private static final String VARIABLE_VALUE = "Joe";

    @Test
    public void should_return_this_object() throws Exception {
        Object data = VARIABLE_VALUE;
        DefaultScope scope = new DefaultScope(data);
        assertEquals(VARIABLE_VALUE, scope.getVariableValue("."));
    }

    @Test
    public void should_return_value_from_map() throws Exception {
        Object data = Collections.singletonMap(VARIABLE_NAME, VARIABLE_VALUE);
        DefaultScope scope = new DefaultScope(data);
        assertEquals(VARIABLE_VALUE, scope.getVariableValue(VARIABLE_NAME));
    }

    @Test
    public void should_return_method_return_value() throws Exception {
        Object data = new Object() {
            @SuppressWarnings("unused")
            public String firstName() {
                return VARIABLE_VALUE;
            }
        };
        DefaultScope scope = new DefaultScope(data);
        assertEquals(VARIABLE_VALUE, scope.getVariableValue(VARIABLE_NAME));
    }

    @Test
    public void should_return_property_value() throws Exception {
        Object data = new Object() {
            @SuppressWarnings("unused")
            public String getFirstName() {
                return VARIABLE_VALUE;
            }
        };
        DefaultScope scope = new DefaultScope(data);
        assertEquals(VARIABLE_VALUE, scope.getVariableValue(VARIABLE_NAME));
    }

    @Test
    public void should_return_field_value() throws Exception {
        Object data = new Object() {
            @SuppressWarnings("unused")
            public String firstName = VARIABLE_VALUE;
        };
        DefaultScope scope = new DefaultScope(data);
        assertEquals(VARIABLE_VALUE, scope.getVariableValue(VARIABLE_NAME));
    }

    @Test
    public void should_return_Callable_return_value() throws Exception {
        Object data = new Object() {
            @SuppressWarnings("unused")
            public Callable<String> firstName = new Callable<String>() {
                public String call() throws Exception {
                    return VARIABLE_VALUE;
                }
            };
        };
        DefaultScope scope = new DefaultScope(data);
        assertEquals(VARIABLE_VALUE, scope.getVariableValue(VARIABLE_NAME));
    }
}
